### PR TITLE
Opensuse support

### PR DESCRIPTION
**Pull request by marcmolla**
## Description of change

This change add the support of OpenSUSE Leap (42 series) to Juju. 

Main reason for the change is that most of our NVF run on top of OpenSUSE/SLES and I was doing some hands-on for evaluating Juju as a generic VNF-M. This PR  (and other for juju/juju) is the result of my work.

With these PR, Juju users are able to deploy OpenSUSE charms in LXD and in manual provisioned OpenSUSE host.

## QA steps

A part of the testing that I added, it is possible to deploy an OpenSUSE charm:
https://github.com/marcmolla/juju-OpenSUSE/tree/master/charms/test-opensuseleap

using a modified OpenSUSE image:
https://github.com/marcmolla/juju-OpenSUSE/blob/master/lxd-opensuse-42.2-image.md

I also describe how I tested the software locally:
https://github.com/marcmolla/juju-OpenSUSE/blob/master/test-compiled-Juju.md

For all those testing we also need another PR to juju/juju that I am going to create -> https://github.com/juju/juju/pull/7257

## Documentation changes

If this PR is accepted, we should include reference to OpenSUSE in general doc.

## Bug reference

N/A

### CODE REVIEW

File: packaging/commands/zypper.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// Created from yum.go
+
+package commands
+
+const (
+	// OpenSUSESourcesDir is the default directory in which openSUSE sourcefiles
+	// may be found.
+	OpenSUSESourcesDir = "/etc/zypp/repos.d"
+
+	// TODO: Something similar in zypper?
```
Comment (axw) [inline packaging/commands/zypper.go]: it's not used anyway, so just drop this one

File: packaging/commands/zypper.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	OpenSUSEProxy = "/etc/sysconfig/proxy"
+)
+
+const (
+	// Zypper command for managing packages and repos
+	//		--quiet to only show errors
+	//		--non-interactive to install without asking packages
+	zypper = "zypper --quiet --non-interactive"
+
+	// TODO: is useful?
```
Comment (axw) [inline packaging/commands/zypper.go]: I'd like to switch proxySettingsFormat to per-scheme strings (i.e. httpProxySettingsFormat, httpsProxySettingsFormat, ftpProxySettingsFormat). Until then, it's fine to have this as a constant; feel free to drop the TODO.

File: packaging/config/opensuse_constants.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+// Copied from centos_constants.go (with all pending TODOs)
+
+package config
+
+import (
+	"github.com/juju/utils/packaging"
+)
+
+const (
+	// Make sense for OpenSUSE?
```
Comment (axw) [inline packaging/config/opensuse_constants.go]: just delete this constant, it's not needed

File: packaging/config/opensuse_constants.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	// CentOSCloudArchiveUrl is the url of the (future) cloud archive for CentOS.
+	// TODO (people of the distant future): add this when it is available.
+	// CentOSCloudArchiveUrl = "http://fill-me-in.com/cloud-archive.repo"
+
+	// CentOSSourcesFile is the default file which lists all core sources
```
Comment (axw) [inline packaging/config/opensuse_constants.go]: s/CentOS/OpenSUSE/

File: packaging/config/opensuse_constants.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	// for yum packages on CentOS.
```
Comment (axw) [inline packaging/config/opensuse_constants.go]: s/yum/zypp/ ?

File: packaging/config/opensuse_constants.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	OpenSUSESourcesFile = "/etc/zypp/repos.d/repo-oss.repo"
+
+)
+
+// CentOSDefaultPackages is the default package set we'd like installed
+// on all CentOS machines.
+var OpenSUSEDefaultPackages = append(DefaultPackages, []string{
+	// TODO (aznashwan, all): further populate this list.
```
Comment (axw) [inline packaging/config/opensuse_constants.go]: unless there are other things required to get Juju working, drop this TODO

File: packaging/config/zypper_constants.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"text/template"
+)
+
+const (
+	// ZypperSourcesDir is the default directory in which yum sourcefiles are located.
+	ZypperSourcesDir = "/etc/zypp/repos.d"
+
+	// YumKeyfileDir is the default directory for yum repository keys.
```
Comment (axw) [inline packaging/config/zypper_constants.go]: just drop this until we know we need it, it can be added in later if necessary

**Comment by marcmolla** [comment]
Thanks for your comments! I updated the PR

File: packaging/commands/commander.go [hunk 8]
```diff
@@ -132,9 +133,21 @@ func (p *packageCommander) proxyConfigLines(settings proxy.Settings) []string {
...
-	addOption("https", settings.Https)
-	addOption("ftp", settings.Ftp)
+	// OpenSUSE uses proxy labels in capital letter (e.g HTTP_PROXY)
+	// For backward compatibility I included a flag in packageCommander.
+	var http_label, https_label, ftp_label string
+	if p.proxyLabelInCapital{
+		http_label = "HTTP"
+		https_label = "HTTPS"
+		ftp_label = "FTP"
+	}else{
```
Comment (axw) [inline packaging/commands/commander.go]: There's some non-standard formatting in this file, please run "go fmt" over all files. (I would do it myself after, but I think the merge bot will reject the merge if the files aren't formatted.)

Comment (marcmolla) [inline packaging/commands/commander.go]: Sorry for that, I used a script for the go fmt that obviously failed.. I am running it manually

**Comment by axw** [comment]
@marcmolla can you please squash your commits, as described in http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by marcmolla** [comment]
Yes, I will try it because I am doing something wrong (I am in a loop with the squashing :))

**Comment by marcmolla** [comment]
It seems that now I dit it, but I have to solve this conflict:

Merge branch 'master' into opensuse_support

not sure if I resolve it correctly

**Comment by marcmolla** [comment]
No, it is not correctly solved. I am on it

File: proxy/proxy.go [hunk 9]
```diff
@@ -133,3 +133,29 @@ func (s *Settings) FullNoProxy() string {
 	noProxySet := set.NewStrings(allNoProxy...)
 	return strings.Join(noProxySet.SortedValues(), ",")
 }
+
+// FullNoProxy merges NoProxy and AutoNoProxyList
```
Comment (axw) [inline proxy/proxy.go]: not sure what happened here, but these two are copies of the function above - please drop the dupes

**Comment by marcmolla** [comment]
not sure what happened, some conflicts of commits that I pulled from the repo

is ok now?

**Comment by axw** [comment]
> is ok now?

No, there is still some other stuff in the commit. I'll see if I can cherry pick your commit into a new branch on Monday, if you don't resolve it before hand.

**Comment by marcmolla** [comment]
ok, this is my best... I am not able to squash these two commits without introducing new changes.

**Comment by axw** [comment]
@marcmolla I've merged your changes in #277. Thanks again! I'll try to review your juju/juju PR in a couple of days. It's a public holiday tomorrow in AU and NZ, so there won't be many people around until Wednesday.

**Pull request closed without merging**